lib/ex_image_info/types/isobmff.ex

```diff
@@ -0,0 +1,313 @@
+      state.readable_bytes == nil || state.readable_bytes > 0
+    end
+
+    def update_readable_bytes(state, size) when size < 0 do
+      if state.readable_bytes == nil do
+        state
+      else
+        %{state | readable_bytes: state.readable_bytes + size}
+      end
+    end
+
+    def update_readable_bytes(state, readable_bytes),
+      do: %{state | readable_bytes: readable_bytes}
+
+    def update_index(state, value \\ nil) do
+      value = if value == nil, do: state.index + 1, else: value
+      %{state | index: value}
+    end
+
+    def ispe_box(state, width, height) do
+      %{state | ispe_boxes: [{state.index, {width, height}} | state.ispe_boxes]}
+    end
+
+    def ipma_box(state, id, property_idx) do
+      %{state | ipma_boxes: [{id, property_idx} | state.ipma_boxes]}
+    end
+  end
+
+  @doc """
+  Parses the binary to get the image size (ISOBMFF spec)
+  """
+  def parse_image_size(<<bin::binary>>), do: parse_box(bin, %State{})
+
+  defp parse_box(<<rest::binary>>, state) do
+    if State.readable_bytes?(state) do
+      case read_box_header(rest, state) do
+        {type, size, rest, state} -> handle_box(type, size, rest, state)
+        malformed -> malformed
+      end
+    else
+      @malformed
+    end
+  end
+
+  defp read_box_header(
+         <<0x01::size(32), type::binary-size(4), extended_size::size(64),
+           rest::binary>>,
+         state
+       ) do
+    # 4 bytes for size field, 4 bytes for type field, 8 bytes for extended size
+    header_size = 4 + 4 + 8
+    state = State.update_readable_bytes(state, -header_size)
+
+    box_size = extended_size - header_size
+    {type, box_size, rest, state}
+  end
+
+  defp read_box_header(<<size::size(32), type::binary-size(4), rest::binary>>, state) do
+    # 4 bytes for size field, 4 bytes for type field
+    header_size = 4 + 4
+    state = State.update_readable_bytes(state, -header_size)
+
+    box_size = size - header_size
+    {type, box_size, rest, state}
+  end
+
+  defp read_box_header(_malformed_binary, _state), do: @malformed
+
+  defp handle_box("ftyp", size, <<rest::binary>>, state) do
+    case rest do
+      <<_::binary-size(size), rest::binary>> ->
+        state =
+          state
+          |> State.update_readable_bytes(-size)
+          |> State.update_index()
+
+        parse_box(rest, state)
+
+      _ ->
+        @malformed
+    end
+  end
+
+  defp handle_box("meta", size, <<_rest::binary>>, _state) when size < 4,
+    do: @unimplemented
+
+  defp handle_box("meta", size, <<_::binary-size(4), rest::binary>>, state) do
+    max = size - 4
+
+    state =
+      state
+      |> State.update_readable_bytes(max)
+      |> State.update_index(0)
+
+    with %State{primary_box: pbox} = state <- parse_box(rest, state),
+         # not pitm box
+         false <- is_nil(pbox) do
+      primary_indices =
+        state.ipma_boxes
+        |> Enum.filter(fn {id, _} -> id == pbox end)
+        |> Enum.map(&elem(&1, 1))
```

Comment:
Why not use `Stream.filter() |> Stream.map()` here instead? This way the index list is only created once?

Reply:
I've been leaning towards over subscribing on CPU rather than memory because you can scale a cluster easier based on cpu utilization without hard crashes like an OOM would cause 😬